src/frontend/src/expr/type_inference/func.rs

```diff
@@ -157,6 +164,54 @@ fn infer_type_for_special(
                 _ => Ok(None),
             }
         }
+        ExprType::Equal
+        | ExprType::NotEqual
+        | ExprType::LessThan
+        | ExprType::LessThanOrEqual
+        | ExprType::GreaterThan
+        | ExprType::GreaterThanOrEqual
+        | ExprType::IsDistinctFrom
+        | ExprType::IsNotDistinctFrom => {
+            ensure_arity!("cmp", | inputs | == 2);
+            match (inputs[0].is_unknown(), inputs[1].is_unknown()) {
+                // `'a' = null` handled by general rules later
+                (true, true) => return Ok(None),
+                // `null = array[1]` where null should have same type as right side
+                // `null = 1` can use the general rule, but return `Ok(None)` here is less readable
+                (true, false) => {
+                    let owned = std::mem::replace(&mut inputs[0], ExprImpl::literal_bool(true));
+                    inputs[0] = owned.cast_implicit(inputs[1].return_type())?;
+                    return Ok(Some(DataType::Boolean));
+                }
+                (false, true) => {
+                    let owned = std::mem::replace(&mut inputs[1], ExprImpl::literal_bool(true));
+                    inputs[1] = owned.cast_implicit(inputs[0].return_type())?;
+                    return Ok(Some(DataType::Boolean));
+                }
+                // Types of both sides are known. Continue.
+                (false, false) => {}
+            }
+            let ok = match (inputs[0].return_type(), inputs[1].return_type()) {
+                // TODO(#3692): handle `Struct`
+                // It should tolerate field name differences and allow castable types.
+                // `row(int, date) = row(bigint, timestamp)`
+
+                // However, PostgreSQL disallows `int[] = bigint[]`
```

Comment:
`However` is contrasting PostgreSQL's behavior between row/struct and array/list, rather than contrasting between PostgreSQL and us. I will update the wording to make it clear.